Autofix SIM102 (NestedIfStatements)

andersk opened the pull request:
I tried to write an autofixer for SIM102 (`NestedIfStatements`) using LibCST to preserve comments. But I ran into two issues that potentially point to general problems with the with the `SourceCodeLocator` API and/or the `Fix::replacement` API. So I’m opening this draft to discuss what the plan should be.

---

Firstly, if the input `if` statement is indented, the output `if` body will be misindented:

```diff
 while 1:
     while 2:
-        if 3:
-            if 4:
-                5
+        if 3 and 4:
+    5
```

In this case, `SourceCodeLocator::slice_source_code_range` fails to strip the indentation from the second and following lines:

```python
if 3:
            if 4:
                5
```

and `Fix::replacement` fails to add that indentation.

---

Secondly, if the input `if` statement is actually an `elif`, the fix fails:

```python
while 1:
    while 2:
        if 3:
            4
        elif 5:
            if 6:
                7
```

In this case, `SourceCodeLocator::slice_source_code_range` returns this invalid Python that LibCST can’t parse:

```python
elif 5:
            if 6:
                7
```

charliermarsh commented:
The latter issue came up in a slightly different form here: https://github.com/charliermarsh/ruff/pull/1694.

charliermarsh commented:
@andersk - Ok, I think this works? We do skip a few cases: (1) if there are any comments _between_ the two `if` statements, those get destroyed in the fix, so we avoid applying the patch in that case (we _do_ preserve comments in the nested `if` body, so that's fine); and (2) if the resulting statement is too long, we avoid applying it.

charliermarsh commented:
Let me know if you have feedback, and how you feel about merging this change.

andersk commented on ruff_cli/src/diagnostics.rs:
```diff
@@ -42,6 +42,7 @@ pub fn lint_path(
     cache: flags::Cache,
     autofix: fix::FixMode,
 ) -> Result<Diagnostics> {
+    println!("{:?}", path.to_string_lossy());
```
Spurious print.

andersk commented on src/rules/flake8_simplify/rules/fix_if.rs:
```diff
@@ -0,0 +1,109 @@
+    let is_elif = contents.starts_with("elif");
+    let module_text = if is_elif { &contents[2..] } else { &contents };
+
+    // PArse Parse the CST.
```
```suggestion
    // Parse the CST.
```

andersk commented on src/rules/flake8_simplify/rules/fix_if.rs:
```diff
@@ -0,0 +1,109 @@
+        bail!("Unable to fix multiline statements");
+    }
+
+    let contents = textwrap::dedent(&contents);
```
This has various failure modes.

```python
while True:
    if True:
        if True:
            """this
is valid"""

            """the indentation on
            this line is significant"""

            "this is" \
"allowed too"

            ("so is"
"this for some reason")
```

charliermarsh replied:
Right, ok, good call. I believe I can fix this by leveraging LibCST's `indentation` field.

charliermarsh replied:
Alright, I think the new version works, but it requires a bunch of trickery.

If the statement is indented, we embed it in a dummy function, like:

```py
def f():
  if ...
```

Then, at the end, we stripe the function. I _think_ this is the most robust way to preserve indentation. It lets us avoid having to do any indentation or dedentation ourselves (but keeps the source code valid for LibCST).

Separately, we _also_ have to handle `elif` blocks via special-case.

andersk commented:
Yeah, that gets the job done I guess.

I cleaned this up a bit more and split out the helper function changes into a series of logical commits suitable for merging with **Rebase and merge**.
